File: v4/rust/iprange-livedb/src/recovery/construction.rs
//! Shared recovery output envelope and finalization.

use crate::contract::{MetaV4, ValueKind};
use crate::error::{Error, Result};
use crate::immutable_output::{Builder, Finished};

use super::page_set::PageSet;
use super::range_build::{write_metadata, BuildResult};
use super::report::{RecoverySink, Reporter};
use super::{RecoveryReport, ScratchCleanup};

#[derive(Debug)]
pub(crate) struct Construction {
    pub(crate) finished: Finished,
    pub(crate) report: RecoveryReport,
    pub(crate) scratch: Option<ScratchCleanup>,
}

#[derive(Debug)]
pub(crate) struct Failure {
    pub(crate) builder: Builder,
    pub(crate) cause: Error,
    pub(crate) report: RecoveryReport,
    pub(crate) scratch: Option<ScratchCleanup>,
}

pub(crate) struct AnalysisFailure {
    pub(crate) cause: Error,
    pub(crate) report: RecoveryReport,
    pub(crate) scratch: Option<ScratchCleanup>,
}

#[allow(clippy::result_large_err)]
pub(crate) fn prepare<A>(
    builder: Builder,
    source: MetaV4,
    kind: ValueKind,
    analyze: impl FnOnce() -> std::result::Result<A, AnalysisFailure>,
) -> std::result::Result<(Builder, A), Failure> {
    if let Err(cause) = require_builder(&builder, source, kind) {
        return Err(failure(builder, cause, RecoveryReport::default(), None));
    }
    match analyze() {
        Ok(analysis) => Ok((builder, analysis)),
        Err(error) => Err(failure(builder, error.cause, error.report, error.scratch)),
    }
}

pub(crate) fn require_builder(builder: &Builder, source: MetaV4, kind: ValueKind) -> Result<()> {
    let output = builder.meta();
    let feed_index_limit = match kind {
        ValueKind::Direct => 0,
        ValueKind::Membership => source.feed_index_limit,
        ValueKind::Structured => source.feed_index_limit,
    };
    if output.address_family != source.address_family
        || output.value_kind != kind
        || output.structure_kind() != source.structure_kind()
        || output.value_tag != source.value_tag
        || output.feed_index_limit != feed_index_limit
        || output.txn_id != 1
    {
        return Err(Error::InvalidArgument(match kind {
            ValueKind::Direct => "recovery output does not match its direct source",
            ValueKind::Membership => "recovery output does not match its membership source",
            ValueKind::Structured => "recovery output does not match its structured source",
        }));
    }
    Ok(())
}

#[allow(clippy::result_large_err)]
pub(crate) fn finish(
    mut builder: Builder,
    metadata: Option<&[u8]>,
    max_heap_bytes: u64,
    retained_heap_bytes: u64,
    report: RecoveryReport,
    scratch: Option<ScratchCleanup>,
) -> std::result::Result<Construction, Failure> {
    if let Err(cause) = write_metadata(&mut builder, metadata, max_heap_bytes, retained_heap_bytes)
    {
        return Err(failure(builder, cause, report, scratch));
    }
    match builder.finish_owned() {
        Ok(finished) => Ok(Construction {
            finished,
            report,
            scratch,
        }),
        Err(error) => Err(failure(error.builder, error.cause, report, scratch)),
    }
}

#[allow(clippy::result_large_err)]
pub(crate) fn complete_ranges<S: RecoverySink>(
    mut builder: Builder,
    metadata: Option<&[u8]>,
    max_heap_bytes: u64,
    retained_heap_bytes: u64,
    report: RecoveryReport,
    sink: &mut S,
    build: impl FnOnce(&mut Builder, &mut Reporter<'_, S>) -> BuildResult,
) -> std::result::Result<Construction, Failure> {
    let mut reporter = Reporter::resume(report, sink);
    let result = build(&mut builder, &mut reporter);
    let report = reporter.finish();
    let scratch = match result {
        Ok(scratch) => scratch,
        Err(error) => return Err(failure(builder, error.cause, report, error.scratch)),
    };
    finish(
        builder,
        metadata,
        max_heap_bytes,
        retained_heap_bytes,
        report,
        scratch,
    )
}

pub(crate) fn failure(
    builder: Builder,
    cause: Error,
    report: RecoveryReport,
    scratch: Option<ScratchCleanup>,
) -> Failure {
    Failure {
        builder,
        cause,
        report,
        scratch,
    }
}

pub(crate) fn analysis_failure(
    cause: Error,
    report: RecoveryReport,
    scratch: Option<ScratchCleanup>,
) -> AnalysisFailure {
    AnalysisFailure {
        cause,
        report,
        scratch,
    }
}

pub(crate) fn analysis_failure_with_pages(
    pages: PageSet,
    cause: Error,
    report: RecoveryReport,
) -> AnalysisFailure {
    match pages.finish(Err(cause)) {
        Err(failure) => analysis_failure(failure.cause, report, failure.cleanup),
        Ok(_) => unreachable!("failed analysis cannot finish successfully"),
    }
}

File: v4/go/internal/recovery/construction.go
package recovery

// Shared recovery output envelope (Rust recovery/construction.rs): the
// destination builder is proven against the source generation before
// the analysis runs, and every failing arm carries the builder, the
// partial report, and the scratch facts for the terminal. The Go peer
// composes the writer draft as the Rust immutable_output Builder.

import (
	"github.com/firehol/iprange/v4/go/internal/format"
	"github.com/firehol/iprange/v4/go/internal/writer"
)

// constructionFailure is one failed construction (Rust
// construction::Failure): the retained builder, the cause, the
// partial report, and the scratch facts.
type constructionFailure struct {
	builder *writer.Draft
	cause   error
	report  RecoveryReport
	scratch any
}

// analysisFailure is one failed analysis (Rust
// construction::AnalysisFailure).
type analysisFailure struct {
	cause   error
	report  RecoveryReport
	scratch any
}

// prepareConstruction proves the builder against the source and runs
// the analysis (Rust construction::prepare: the builder refusal is the
// invalid-argument class of the source kind).
func prepareConstruction(builder *writer.Draft, source format.Meta, kind uint8, analyze func() (any, *analysisFailure)) (any, *constructionFailure) {
	if err := requireBuilder(builder, source, kind); err != nil {
		return nil, constructionFailureOf(builder, err, RecoveryReport{}, nil)
	}
	analysis, err := analyze()
	if err != nil {
		return nil, constructionFailureOf(builder, err.cause, err.report, err.scratch)
	}
	return analysis, nil
}

// requireBuilder proves the destination builder matches the source
// generation and the recovery kind (Rust construction::require_builder:
// the family, the tag, the feed index limit, and the recovery
// starting transaction).
func requireBuilder(builder *writer.Draft, source format.Meta, kind uint8) error {
	output := builder.Meta()
	purpose := "recovery output does not match its source"
	switch kind {
	case format.ValueKindDirect:
		purpose = "recovery output does not match its direct source"
	case format.ValueKindMembership:
		purpose = "recovery output does not match its membership source"
	case format.ValueKindStructured:
		purpose = "recovery output does not match its structured source"
	}
	feedIndexLimit := uint64(0)
	if kind != format.ValueKindDirect {
		feedIndexLimit = source.FeedIndexLimit
	}
	if output.AddressFamily != source.AddressFamily ||
		output.ValueKind != kind ||
		output.StructureKind != source.StructureKind ||
		output.ValueTag != source.ValueTag ||
		output.FeedIndexLimit != feedIndexLimit ||
		output.TxnID != 1 {
		return &format.Error{Code: format.CodeInvalidArgument, Detail: purpose}
	}
	return nil
}

// constructionFailureOf builds one failed construction (Rust
// construction::failure).
func constructionFailureOf(builder *writer.Draft, cause error, report RecoveryReport, scratch any) *constructionFailure {
	return &constructionFailure{builder: builder, cause: cause, report: report, scratch: scratch}
}

// analysisFailureOf builds one failed analysis (Rust
// construction::analysis_failure).
func analysisFailureOf(cause error, report RecoveryReport, scratch any) *analysisFailure {
	return &analysisFailure{cause: cause, report: report, scratch: scratch}
}

// analysisFailureWithPages folds the page set terminal into one failed
// analysis (Rust construction::analysis_failure_with_pages).
func analysisFailureWithPages(pages *pageSet, cause error, report RecoveryReport) *analysisFailure {
	failure := pages.finish(cause)
	return analysisFailureOf(failure.cause, report, failure.cleanup)
}
